andrepereiradasilva/joomla-cms
# Cookies review

Pull Request for Improvement.
#### Summary of Changes
1. Use epoch + 1 to delete the cookie. I think it's not necessary time() - xxxx. Epoch + 1 should be enough. Also we can have problems with timezones if not putting time() less a full day.
2. Use always '' (empty) when deleting a cookie.
3. Use the Api when deleting the session cookie.
4. Always use HTTPOnly for cookies so they can't be accessed through JS.
5. If HTTPS is forced in the client (Site/Admin) always use only secure cookies
6. Don't use `app->isSSLConnection()` for checking if the cookie should be secure. We could have problem setting a secure cookie in a dual HTTP/HTTPS site. Use the new `$app->isHttpsForced()` instead.
7. Don't use a new language cookie if the cookie language is defined in the language filter plugin to be only session. Use the session instead (we already have a cookie...).

And minor code and code style improvements.
#### Testing Instructions

Since this impacts the users/admin sessions the tests must be made with special attention.

First do a code review.

Then, and since this deals with session cookies, you should apply this patch on latest staging and  watch the cookies that are created in the developer console (Resources -> Cookies) in the following processes:
- site login/logout process
- backend login/logout process
- the remember me process in site authentication
- in a multilanguage site, the remember of the user language (aka the language cookie)
#### Observarions

I checked all the cookies in joomla core (i think i'm not missing none), one i don't know what it does (banner tracks), but i also updated that one.
